### PR TITLE
Zachmu/sql correctness

### PR DESCRIPTION
**Pull request by zachmu**
Fixed two bugs: NULL semantics in IN expressions, and backward indexes for Ascend / Descend indexes on multi-column indexes.

### CODE REVIEW

**Comment by zachmu** [comment]
I also added result caching to subquery expressions, forgot that in the original PR description

File: enginetest/script_queries.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Liquidata, Inc.
```
Comment (Hydrocharged) [inline enginetest/script_queries.go:1]: I like this!

File: sql/expression/in.go [hunk 2]
```diff
@@ -62,9 +59,15 @@ func (in *InTuple) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 	}
 
 	if left == nil {
-		return nil, err
+		return nil, nil
 	}
 
+	// The NULL handling for IN expressions is tricky. According to
```
Comment (Hydrocharged) [inline sql/expression/in.go:65]: This threw me off at first when I was looking at the tests earlier in the PR, and had to look it up. This stuff is not straightforward lol

File: sql/analyzer/resolve_subqueries.go [hunk 3]
```diff
@@ -57,3 +58,33 @@ func resolveSubqueryExpressions(ctx *sql.Context, a *Analyzer, n sql.Node, scope
 		return s.WithQuery(analyzed), nil
 	})
 }
+
+// cacheSubqueryResults determines whether it's safe to cache the results for any subquery expressions. Caching
+// subquery results is safe in the case that no outer scope columns are referenced.
+func cacheSubqueryResults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
```
Comment (Hydrocharged) [inline sql/analyzer/resolve_subqueries.go:64]: Perhaps this isn't an issue and I'm completely off base, but wouldn't this cache a nondeterministic subquery such as `(SELECT FLOOR(RAND()*5) IN (1, 2, 3)`?

Comment (zachmu) [reply sql/analyzer/resolve_subqueries.go:64]: No, you are totally correct. I'll add a TODO for this, nice catch.

Comment (zachmu) [reply sql/analyzer/resolve_subqueries.go:65]: I went to fix this and add tests, and discovered that result caching was totally broken, so thanks for this!

**Pull request merged**